lib/elixir/lib/calendar.ex

```diff
@@ -796,6 +710,75 @@ defmodule NaiveDateTime do
   end
 
   @doc """
+  Adds a specified amount of time to a `NaiveDateTime`.
+
+  Accepts an `integer` in any `unit` available from `System.time_unit`.
+  Negative values will be move backwards in time.
+
+  ## Examples
+
+      # adds seconds by default
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], 2)
+      ~N[2014-10-02 00:29:12]
+      # accepts negative offsets
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], -2)
+      ~N[2014-10-02 00:29:08]
+      # can work with other units
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], 2_000, :milliseconds)
+      ~N[2014-10-02 00:29:12]
+      # doesn't lose precision
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10.021], 21, :milliseconds)
+      ~N[2014-10-02 00:29:10.042]
+      # from gregorian seconds
+      iex> NaiveDateTime.add(~N[0000-01-01 00:00:00], 63579428950)
+      ~N[2014-10-02 00:29:10]
+  """
+  @spec add(NaiveDateTime.t, integer, System.time_unit) :: NaiveDateTime.t
+  def add(%NaiveDateTime{} = naive_datetime,
+          integer,
+          unit \\ :seconds) when is_integer(integer) do
+    ndt_microseconds = to_microseconds(naive_datetime)
+    added_microseconds = System.convert_time_unit(integer, unit, :microseconds)
+    sum = ndt_microseconds + added_microseconds
+
+    microsecond = rem(sum, 1_000_000)
+    precision =
+      if microsecond == 0, do: 0, else: Calendar.ISO.precision_for_unit(unit)
+    {{year, month, day}, {hour, minute, second}} =
+      sum |> div(1_000_000) |> :calendar.gregorian_seconds_to_datetime
+    %NaiveDateTime{year: year, month: month, day: day,
+                   hour: hour, minute: minute, second: second,
+                   microsecond: {microsecond, precision}}
+  end
+
+  @doc """
+  Subtract `naive_datetime_2` from `naive_datetime_1`.
+
+  The answer can be returned in any `unit` available from `System.time_unit`.
+
+  ## Examples
+
+      iex> NaiveDateTime.diff(~N[2014-10-02 00:29:12], ~N[2014-10-02 00:29:10])
+      2
+      iex> NaiveDateTime.diff(~N[2014-10-02 00:29:12], ~N[2014-10-02 00:29:10], :microsecond)
+      2_000_000
+      iex> NaiveDateTime.diff(~N[2014-10-02 00:29:10.042], ~N[2014-10-02 00:29:10.021], :millisecond)
+      21
+      # to gregorian seconds
+      iex> NaiveDateTime.diff(~N[2014-10-02 00:29:10], ~N[0000-01-01 00:00:00])
+      63579428950
+  """
+  @spec diff(NaiveDateTime.t, NaiveDateTime.t, System.time_unit) :: integer
+  def diff(%NaiveDateTime{} = naive_datetime_1,
+           %NaiveDateTime{} = naive_datetime_2,
+           unit \\ :seconds) do
```

Comment:
Please do double-check me.  I tried to catch them all, but I may have missed something.